lib/awaitableSender.ts

```diff
@@ -0,0 +1,178 @@
+
+    const onSuccess = (delivery: Delivery) => {
+      const id = delivery.id;
+      if (this.deliveryDispositionMap.has(delivery.id)) {
+        const promise = this.deliveryDispositionMap.get(id) as PromiseLike;
+        clearTimeout(promise.timer);
+        const deleteResult = this.deliveryDispositionMap.delete(id);
+        log.sender(
+          "[%s] Event: 'Accepted', Successfully deleted the delivery with id %d from " +
+          "the map of sender '%s' on amqp session '%s' and cleared the timer: %s.",
+          this.connection.id, id, this.name, this.session.id, deleteResult
+        );
+        return promise.resolve(delivery);
+      }
+    };
+
+    const onFailure = (eventName: string, id: number, error?: Error) => {
+      if (this.deliveryDispositionMap.has(id)) {
+        const promise = this.deliveryDispositionMap.get(id) as PromiseLike;
+        clearTimeout(promise.timer);
+        const deleteResult = this.deliveryDispositionMap.delete(id);
+        log.sender(
+          "[%s] Event: '%s', Successfully deleted the delivery with id %d from the " +
+          " map of sender '%s' on amqp session '%s' and cleared the timer: %s.",
+          this.connection.id, eventName, id, this.name, this.session.id, deleteResult
+        );
+        const msg = `Sender '${this.name}' on amqp session '${this.session.id}', received a ` +
+          `'${eventName}' disposition. Hence we are rejecting the promise.`;
+        const err = new SendOperationFailedError(msg, error);
+        log.error("[%s] %s", this.connection.id, msg);
+        return promise.reject(err);
+      }
+    };
+
+    const onError = (eventName: string, error?: Error) => {
+      for (const id of this.deliveryDispositionMap.keys()) {
+        onFailure(eventName, id, error);
+      }
+    };
+
+    this.on(SenderEvents.accepted, (context: EventContext) => {
+      onSuccess(context.delivery!);
+    });
+    this.on(SenderEvents.rejected, (context: EventContext) => {
+      const delivery = context.delivery!;
+      onFailure(SenderEvents.rejected, delivery.id, delivery.remote_state && delivery.remote_state.error);
+    });
+    this.on(SenderEvents.released, (context: EventContext) => {
+      const delivery = context.delivery!;
+      onFailure(SenderEvents.released, delivery.id, delivery.remote_state && delivery.remote_state.error);
+    });
+    this.on(SenderEvents.modified, (context: EventContext) => {
+      const delivery = context.delivery!;
+      onFailure(SenderEvents.modified, delivery.id, delivery.remote_state && delivery.remote_state.error);
+    });
+
+    // The user may have it's custom reconnect logic for bringing the sender link back online and
+    // retry logic for sending messages on failures hence they can provide their error handlers
+    // for sender_error and session_error.
+    // If the user did not provide their error handler for sender_error and session_error, then we
```

Comment:
If user does provide their own error handlers, then there would be a case where the `deliveryDispositionMap` might not get cleared when long timeouts are given right? Maybe, we don't care about this much, just a thought.

Reply:
Hence `deliveryDispositionMap` is kept public. So the user can access it and process the entries in the map as per its liking. Will add a line about this in the documentation.